fasthash-sys/build.rs

```diff
@@ -156,6 +156,9 @@ fn build_t1() {
         .file("src/t1ha/src/t1ha1.c")
         .file("src/t1ha/src/t1ha2.c");
 
+    // indirect functions are not supported on all targets (e.g. x86_64-unknown-linux-musl)
+    build.define("T1HA_USE_INDIRECT_FUNCTIONS", Some("0"));
```

Comment:
This doesn't even compile with this flag for me. ```undefined reference to `t1ha0'```.

Reply:
You're right, `cargo test --target=x86_64-unknown-linux-musl` currently fails (tested on macOS and Ubuntu VM), but the flag does fix cross compilation cargo builds on macOS.
The tests still pass on `x86_64-apple-darwin` and `x86_64-unknown-linux-gnu` so I didn't notice (I only use musl-libc for release builds).

Even without the `T1HA_USE_INDIRECT_FUNCTIONS` build flag the test compilation fails for me due to smhasher:
```
/usr/bin/ld: /home/ubuntuvm/Desktop/rust-fasthash/target/x86_64-unknown-linux-musl/debug/deps/libfasthash_sys-e9ce91fe095e797a.rlib(fasthash.o): in function `mum_hash(void const*, unsigned long, unsigned long)':
          /home/ubuntuvm/Desktop/rust-fasthash/fasthash-sys/src/smhasher/mum.h:411: undefined reference to `__cpu_indicator_init'
          /usr/bin/ld: /home/ubuntuvm/Desktop/rust-fasthash/fasthash-sys/src/smhasher/mum.h:412: undefined reference to `__cpu_model'
          collect2: error: ld returned 1 exit status
```

Are you able to compile the tests for `x86_64-unknown-linux-musl` at all when the new flag is commented out?

Reply:
For me `cargo test --target=x86_64-unknown-linux-gnu`  fails.
```
  = note: /usr/bin/ld: rust-fasthash/target/x86_64-unknown-linux-gnu/debug/deps/libfasthash_sys-a0d5e1f8da5f5764.rlib(fasthash.o): in function `t1ha0_64(void const*, unsigned long, unsigned long)':
          rust-fasthash/fasthash-sys/src/fasthash.cpp:60: undefined reference to `t1ha0'
          /usr/bin/ld: rust-fasthash/target/x86_64-unknown-linux-gnu/debug/deps/fasthash-0826da2faaa60bb4: hidden symbol `t1ha0' isn't defined
          /usr/bin/ld: final link failed: bad value
          collect2: error: ld returned 1 exit status
```
and it does work without this line. I had to pin the previous commit on the corresponding projects to get it work.

Reply:
@cycraig If this doesn't even fix musl then we can just as well revert this change right? Sounds to me a lot like you were just looking for a way to disable t1ha completely.

Reply:
>> but the flag does fix cross compilation cargo builds on macOS
>
> What does this mean?

I mentioned in the pull request that without the flag cross-compiling to musl on macOS (`cargo build --target=x86_64-unknown-linux-musl`) was failing with the following error:

```
running: "x86_64-linux-musl-gcc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-static" "-Wall" "-Wextra" "-maes" "-mavx" "-mavx2" "-DT1HA0_RUNTIME_SELECT=1" "-DT1HA0_AESNI_AVAILABLE=1" "-o" "/Users/<removed>/target/x86_64-unknown-linux-musl/release/build/fasthash-sys-f1630acf3eefd1ac/out/src/t1ha/src/t1ha0.o" "-c" "src/t1ha/src/t1ha0.c"
cargo:warning=src/t1ha/src/t1ha0.c:434:10: error: ifunc is not supported on this target
cargo:warning=  434 | uint64_t t1ha0(const void *data, size_t len, uint64_t seed)
cargo:warning=      |          ^~~~~
exit code: 1
```
The flag allowed me to compile a working release build targeting musl on macOS.

> Sounds to me a lot like you were just looking for a way to disable t1ha completely for musl.

I can cargo build to musl without disabling the t1ha feature (but I don't call it in my project so I can't say whether or not t1ha works on the target). Making it optional was just good to reduce the library size if unused, it was unrelated to any compilation errors.

It seems to be a case of "works on my machine"TM on both our accounts then. On a clean install of Ubuntu I still can't compile tests for musl even using previous commits --- neither with the default musl-tools from apt nor https://github.com/richfelker/musl-cross-make --- due to the `undefined reference to '__cpu_indicator_init' [...] '__cpu_model'` error I mentioned before.
Fixing that by manually editing fasthash-sys/src/smhasher/mum.h allows the tests to compile but they still fail on musl for me:
```
Finished test [unoptimized + debuginfo] target(s) in 1m 15s
     Running target/x86_64-unknown-linux-musl/debug/deps/fasthash-9291f9ca4de7d7b0
error: test failed, to rerun pass '-p fasthash --lib'

Caused by:
  could not execute process `/home/ubuntu/Desktop/fasthash/target/x86_64-unknown-linux-musl/debug/deps/fasthash-9291f9ca4de7d7b0` (never executed)
```
I get the same result when trying to run the tests on an Alpine VM.

Seeing as I can't really make headway on the issue on any of my local environments and the change breaks tests that were working on your side, you can probably just revert the change or hide it behind a feature.

Reply:
So `#[cfg(target_env = "musl")]` would be alright for you?

Reply:
> The flag allowed me to compile a working release build targeting musl on macOS.

To be honest I don't even  get where the alternative definition of `t1ha0` comes from with this flag set to zero.

Reply:
> So `#[cfg(target_env = "musl")]` would be alright for you? See #20

Oh, I thought we were talking about the tests failing on musl only, I misread the "gnu" when you posted `x86_64-unknown-linux-gnu`, sorry for the confusion. `cargo test` passes on MacOS because t1ha.h sets the flag to 0 when the compiler doesn't support ELF binaries anyway.

This change works fine, thanks.